File: src/lib/html-helpers.ts
// HTML utility functions

export function escapeHtml(str: string | undefined): string {
  if (!str) return '';
  return str.replace(/[&<>"']/g, c => ({
    '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;'
  }[c] || c));
}

export function isValidHexColor(color: string): boolean {
  return /^#[0-9A-Fa-f]{6}$/.test(color);
}

export function sanitizeColor(color: string | undefined, fallback: string): string {
  if (!color) return fallback;
  return isValidHexColor(color) ? color : fallback;
}

export function getOverlayGradient(overlay: string | undefined): string {
  switch (overlay) {
    case 'light': return 'linear-gradient(rgba(0,0,0,0.2), rgba(0,0,0,0.2))';
    case 'medium': return 'linear-gradient(rgba(0,0,0,0.4), rgba(0,0,0,0.4))';
    case 'dark': return 'linear-gradient(rgba(0,0,0,0.6), rgba(0,0,0,0.6))';
    default: return 'none';
  }
}

// Sanitize URL for CSS context - prevent CSS injection via url()
export function sanitizeUrlForCss(url: string | undefined): string {
  if (!url) return '';
  // Remove dangerous characters that could break out of CSS url()
  const sanitized = url.replace(/['"()\\]/g, '');
  // Only allow http/https/data URLs or relative paths
  if (/^(https?:\/\/|data:|\/)/i.test(sanitized)) {
    return sanitized;
  }
  return '';
}

// Map logo_position string to inline CSS position string (default: top-right)
export function logoPositionStyle(pos: string | undefined): string {
  const map: Record<string, string> = {
    'top-left': 'top:5%;left:5%',
    'top-right': 'top:5%;right:5%',
    'bottom-left': 'bottom:5%;left:5%',
    'bottom-right': 'bottom:5%;right:5%',
  };
  return map[pos || ''] || map['top-right'];
}

// Render watermark img overlay or empty string if no URL provided
export function watermarkHtml(watermarkUrl: string | undefined, opacity: string | undefined): string {
  if (!watermarkUrl) return '';
  const safe = sanitizeUrlForCss(watermarkUrl);
  if (!safe) return '';
  const opacityValue = ({ light: '0.2', medium: '0.4', dark: '0.6' } as Record<string, string>)[opacity || 'light'] || '0.2';
  return `<img src="${safe}" alt="" style="position:absolute;bottom:3%;right:3%;height:6%;max-width:20%;object-fit:contain;opacity:${opacityValue};pointer-events:none;z-index:10;" crossorigin="anonymous" onerror="this.style.display='none'" />`;
}

File: server/engine/html_helpers.go
package engine

import (
	"fmt"
	"regexp"
	"strings"
)

var hexColorRe = regexp.MustCompile(`^#[0-9A-Fa-f]{6}$`)

// EscapeHTML escapes HTML special characters
func EscapeHTML(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	s = strings.ReplaceAll(s, `"`, "&quot;")
	s = strings.ReplaceAll(s, "'", "&#39;")
	return s
}

// SanitizeColor returns color if valid hex, else fallback
func SanitizeColor(color, fallback string) string {
	if color == "" {
		return fallback
	}
	if hexColorRe.MatchString(color) {
		return color
	}
	return fallback
}

// SanitizeURLForCSS prevents CSS injection via url()
func SanitizeURLForCSS(url string) string {
	if url == "" {
		return ""
	}
	// Remove dangerous characters
	sanitized := strings.NewReplacer("'", "", `"`, "", "(", "", ")", "", `\`, "").Replace(url)
	// Only allow http/https/data URLs or relative paths
	if strings.HasPrefix(sanitized, "http://") || strings.HasPrefix(sanitized, "https://") ||
		strings.HasPrefix(sanitized, "data:") || strings.HasPrefix(sanitized, "/") {
		return sanitized
	}
	return ""
}

// GetOverlayGradient returns CSS gradient for overlay level
func GetOverlayGradient(overlay string) string {
	switch overlay {
	case "light":
		return "linear-gradient(rgba(0,0,0,0.2), rgba(0,0,0,0.2))"
	case "medium":
		return "linear-gradient(rgba(0,0,0,0.4), rgba(0,0,0,0.4))"
	case "dark":
		return "linear-gradient(rgba(0,0,0,0.6), rgba(0,0,0,0.6))"
	default:
		return "none"
	}
}

// LogoPositionStyle returns inline CSS for logo placement
func LogoPositionStyle(pos string) string {
	m := map[string]string{
		"top-left":     "top:5%;left:5%",
		"top-right":    "top:5%;right:5%",
		"bottom-left":  "bottom:5%;left:5%",
		"bottom-right": "bottom:5%;right:5%",
	}
	if s, ok := m[pos]; ok {
		return s
	}
	return m["top-right"]
}

// WatermarkHTML returns watermark img element or empty string
func WatermarkHTML(wmURL, opacity string) string {
	if wmURL == "" {
		return ""
	}
	safe := SanitizeURLForCSS(wmURL)
	if safe == "" {
		return ""
	}
	opMap := map[string]string{"light": "0.2", "medium": "0.4", "dark": "0.6"}
	opVal := opMap[opacity]
	if opVal == "" {
		opVal = "0.2"
	}
	return fmt.Sprintf(`<img src="%s" alt="" style="position:absolute;bottom:3%%;right:3%%;height:6%%;max-width:20%%;object-fit:contain;opacity:%s;pointer-events:none;z-index:10;" crossorigin="anonymous" onerror="this.style.display='none'" />`, safe, opVal)
}

// HasImage checks if URL is valid and not "undefined"/"null"
func HasImage(url string) bool {
	return url != "" && url != "undefined" && url != "null"
}

// LogoHTML returns logo img element
func LogoHTML(logo, position, height, maxWidth string) string {
	safe := SanitizeURLForCSS(logo)
	if safe == "" {
		return ""
	}
	if height == "" {
		height = "8%"
	}
	if maxWidth == "" {
		maxWidth = "25%"
	}
	return fmt.Sprintf(`<img src="%s" alt="logo" style="position:absolute;%s;height:%s;max-width:%s;object-fit:contain;" crossorigin="anonymous" onerror="this.style.display='none'" />`,
		safe, LogoPositionStyle(position), height, maxWidth)
}
